core/domain/exp_domain.py

```diff
@@ -474,6 +474,285 @@ def __init__(self, change_list):
+        return [self[v] for v in python_utils.RANGE(version_start, version_end)]
+
+    @staticmethod
+    def _default_state_equality_predicate(unused_state1, unused_state2):
+        """Always returns True."""
+        return True
+
+
+class ExplorationStatesHistory(python_utils.OBJECT):
+    """Organizes the history of changes made to each state of an exploration."""
+
+    def __init__(self, exps, exp_version_diffs, state_equality_predicate=None):
+        """Constructs the exploration's state history.
+
+        Args:
+            exps: list(Exploration). The exploration's representation, ordered
+                by version. Items should begin at version 1 and contain no gaps.
+            exp_version_diffs: list(ExplorationVersionsDiff). The specific
+                changes made to the exploration at each corresponding version.
+            state_equality_predicate:
+                callable(state_domain.State, state_domain.State) -> bool | None.
+                Returns True when two states are "equal". The predicate is used
+                to enforce invariant: all states in a span are equivalent.
+                If None, then states will always be equivalent to each other.
+        """
+        if not exps or not exp_version_diffs:
+            raise ValueError('Inputs must be non-empty')
+
+        if len(exps) != len(exp_version_diffs):
+            raise ValueError('len of inputs do not match')
+
+        exp_versions = [exp.version for exp in exps]
+        expected_versions = list(python_utils.RANGE(1, len(exps) + 1))
+        if exp_versions != expected_versions:
+            raise ValueError(
+                'Explorations must be sorted, begin from version 1, and have '
+                'no gaps; but got invalid input versions == [%s]' % (
+                    ', '.join(python_utils.UNICODE(v) for v in exp_versions)))
+
+        self._state_spans_history = []
+        for exp, diff in python_utils.ZIP(exps, exp_version_diffs):
+            new_state_spans = dict()
+            for state_name, state in exp.states.items():
+                if exp.version == 1 or state_name in diff.added_state_names:
+                    new_state_spans[state_name] = StateVersionSpan(
+                        exp.version, state_name, state,
+                        state_equality_predicate=state_equality_predicate)
+                elif state_name not in diff.deleted_state_names:
+                    prev_state_spans, prev_name = (
+                        self._state_spans_history[-1],
+                        diff.new_to_old_state_names.get(state_name, state_name))
+                    new_state_spans[state_name] = (
+                        prev_state_spans[prev_name].extend_or_split(
+                            exp.version, state_name, state, diff))
+            self._state_spans_history.append(new_state_spans)
+
+    def get_state_span(self, exp_version, state_name):
+        """Returns the span of versions where the given state existed.
```

Comment:
Probably this description should have some reference to exp_version (it sounds like it's independent of it).